src/animation.rs

```diff
@@ -0,0 +1,503 @@
+    Scale,
+}
+
+/// An index into the frames of a track.
+enum Frame {
+    /// The time is before the start of the frames.
+    Unstarted,
+
+    /// The time corresponds to the given frame index.
+    InProgress(usize),
+
+    /// The time is after the end of the last frame.
+    Ended,
+}
+
+/// The keyframe values of a [`Track`].
+///
+/// [`Track`]: struct.Track.html
+#[derive(Clone, Debug)]
+pub enum Values {
+    /// Quaternion keyframes.
+    Quaternion(Vec<mint::Quaternion<f32>>),
+
+    /// 3D vector keyframes.
+    Vector3(Vec<mint::Vector3<f32>>),
+
+    /// Scalar keyframes.
+    Scalar(Vec<f32>),
+}
+
+pub(crate) enum Operation {
+    Enable,
+    Disable,
+    Pause,
+    Play,
+    SetLoopMode(LoopMode),
+}
+
+/// Controls the playback properties of an animation
+pub struct Action {
+    /// Message channel to hub.
+    tx: mpsc::Sender<Message>,
+
+    /// Pointer to parent `Mixer` struct.
+    pointer: froggy::Pointer<ActionData>,
+}
+
+/// Internal data for an animation action.
+struct ActionData {
+    /// The animation data for this action.
+    pub clip: Clip,
+
+    /// Specifies whether the action is enabled or disabled.
+    ///
+    /// A disabled action has no impact.
+    pub enabled: bool,
+
+    /// Specifies the looping behaviour of this action.
+    pub loop_mode: LoopMode,
+
+    /// Specifies whether the action is paused.
+    pub paused: bool,
+
+    /// The local time of this action in seconds, starting at 0.0.
+    ///
+    /// This value get clamped or wrapper to [0.0, clip.duration] depending on
+    /// the loop mode.
+    pub local_time: f32,
+
+    /// Time scaling factor.
+    pub local_time_scale: f32,
+
+    // Unimplemented properties
+    // ------------------------
+    // * weight
+    // * zero_slope_at_end
+    // * zero_slope_at_start
+}
+
+/// A reusable set of keyframe tracks which represent an animation.
+#[derive(Clone, Debug)]
+pub struct Clip {
+    /// A name for this clip.
+    pub name: Option<String>,
+
+    /// The animation keyframe tracks.
+    pub tracks: Vec<Track>,
+}
+
+/// A track of animation keyframes.
+#[derive(Clone, Debug)]
+pub struct Track {
+    /// The object this track updates.
+    pub object: Object,
+
+    /// The object property this track updates.
+    pub binding: Binding,
+
+    /// The keyframe time values.
+    pub times: Vec<f32>,
+
+    /// The keyframe values.
+    pub values: Values,
+
+    /// Specifies the interpolation strategy between keyframes.
+    pub interpolation: Interpolation,
+}
+
+type Message = (froggy::WeakPointer<ActionData>, Operation);
+
+/// Scheduler for the playback of animation actions.
+///
+/// Use this to update animation actions.
+pub struct Mixer {
+    actions: froggy::Storage<ActionData>,
+    rx: mpsc::Receiver<Message>,
+    tx: mpsc::Sender<Message>,
+}
+
+impl Action {
+    fn send(
+        &mut self,
+        operation: Operation,
+    ) -> &mut Self {
+        let message = (self.pointer.downgrade(), operation);
```

Comment:
Did I implement this correctly? The receiving end is at `Mixer::update`.

Reply:
looks fine to me